File: LectureHSE1/LectureHSE1/Screens/Home/Model/DTO/NewsResponseDTO.swift
//
//  NewsResponseDTO.swift
//  LectureHSE1
//
//  Created by Zaitsev Vladislav on 05.03.2026.
//

import Foundation

struct NewsResponseDTO: Codable {
    let status: String
    let copyright: String
    let response: NewsSearchResponseDTO
}

struct NewsSearchResponseDTO: Codable {
    let docs: [NewsItemDTO]
    let metadata: NewsMetadataDTO?
}

struct NewsMetadataDTO: Codable {
    let hits: Int?
    let offset: Int?
    let time: Int?
}

struct NewsItemDTO: Codable {
    let id: String?
    let abstract: String?
    let snippet: String?
    let webUrl: String?
    let source: String?
    let pubDate: String?
    let sectionName: String?
    let subsectionName: String?
    let headline: HeadlineDTO?
    let byline: BylineDTO?
    let multimedia: MultimediaDTO?
    let uri: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case abstract
        case snippet
        case webUrl = "web_url"
        case source
        case pubDate = "pub_date"
        case sectionName = "section_name"
        case subsectionName = "subsection_name"
        case headline
        case byline
        case multimedia
        case uri
    }
}

struct HeadlineDTO: Codable {
    let main: String?
    let kicker: String?
    let printHeadline: String?

    enum CodingKeys: String, CodingKey {
        case main
        case kicker
        case printHeadline = "print_headline"
    }
}

struct BylineDTO: Codable {
    let original: String?
}

struct MultimediaDTO: Codable {
    let caption: String?
    let credit: String?
    let defaultImage: MultimediaImageDTO?
    let thumbnail: MultimediaImageDTO?

    enum CodingKeys: String, CodingKey {
        case caption
        case credit
        case defaultImage = "default"
        case thumbnail
    }
}

struct MultimediaImageDTO: Codable {
    let url: String?
    let height: Int?
    let width: Int?
}

File: ProjectAndroid/app/src/main/java/com/example/projectmobileandroid/Home/Model/NewsResponseDTO.kt
package com.example.projectmobileandroid.Home.Model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class NewsResponseDTO(
    val status: String = "",
    val copyright: String = "",
    val response: NewsSearchResponseDTO = NewsSearchResponseDTO()
)

@Serializable
data class NewsSearchResponseDTO(
    val docs: List<NewsItemDTO> = emptyList(),
    val metadata: NewsMetadataDTO = NewsMetadataDTO()
)

@Serializable
data class NewsMetadataDTO(
    val hits: Int = 0,
    val offset: Int = 0,
    val time: Int = 0
)

@Serializable
data class NewsItemDTO(
    @SerialName("_id")
    val id: String? = null,
    val abstract: String? = null,
    val snippet: String? = null,
    @SerialName("web_url")
    val webUrl: String? = null,
    val source: String? = null,
    @SerialName("pub_date")
    val pubDate: String? = null,
    @SerialName("section_name")
    val sectionName: String? = null,
    @SerialName("subsection_name")
    val subsectionName: String? = null,
    val headline: HeadlineDTO? = null,
    val byline: BylineDTO? = null,
    val multimedia: MultimediaDTO? = null,
    val uri: String? = null
)

@Serializable
data class HeadlineDTO(
    val main: String? = null,
    val kicker: String? = null,
    @SerialName("print_headline")
    val printHeadline: String? = null
)

@Serializable
data class BylineDTO(
    val original: String? = null
)

@Serializable
data class MultimediaDTO(
    val caption: String? = null,
    val credit: String? = null,
    @SerialName("default")
    val defaultImage: MultimediaImageDTO? = null,
    val thumbnail: MultimediaImageDTO? = null
)

@Serializable
data class MultimediaImageDTO(
    val url: String? = null,
    val height: Int? = null,
    val width: Int? = null
)
